### PR TITLE
Improve hipcaffe port

### PR DESCRIPTION
**Pull request by whchung**
- Fix one memory leak
- Reduce HIP stream numbers by using default HIP stream
- Reduce MIOpen handle numbers by one handle for fwd / bwd weight / bwd data
- Polish debug message output
- Add debug message for CuDNNConvolutionLayer count

- Remove unused CUDNN integration logic

### CODE REVIEW

**Comment by whchung** [comment]
@sunway513  / @parallelo please help review the PR. Thanks a lot.

File: src/caffe/layers/cudnn_conv_layer.cpp [hunk 1]
```diff
@@ -636,9 +615,15 @@ CuDNNConvolutionLayer<Dtype>::~CuDNNConvolutionLayer() {
   delete [] bwd_filter_algo_;
   delete [] bwd_data_algo_;
 #endif
+  delete [] workspace;
```
Comment (whchung) [inline src/caffe/layers/cudnn_conv_layer.cpp]: This is a memory leak fixed.

File: include/caffe/layers/cudnn_conv_layer.hpp [hunk 2]
```diff
@@ -29,8 +29,12 @@ namespace caffe {
 template <typename Dtype>
 class CuDNNConvolutionLayer : public ConvolutionLayer<Dtype> {
  public:
-  explicit CuDNNConvolutionLayer(const LayerParameter& param)
-      : ConvolutionLayer<Dtype>(param), handles_setup_(false) {}
+#ifdef CONV_LAYER_COUNT
+  static int count;
```
Comment (whchung) [inline include/caffe/layers/cudnn_conv_layer.hpp:33]: debug routines to help count instances of `CuDNNConvolutionLayer`.

File: include/caffe/layers/cudnn_conv_layer.hpp [hunk 3]
```diff
@@ -86,6 +88,7 @@ class CuDNNConvolutionLayer : public ConvolutionLayer<Dtype> {
   size_t workspaceSizeInBytes;  // size of underlying storage
   void *workspaceData;  // underlying storage
   void **workspace;  // aliases into workspaceData
+  miopenHandle_t handle_;
```
Comment (whchung) [inline include/caffe/layers/cudnn_conv_layer.hpp:76]: - remove `stream_` so we only use default HIP stream.
- change `handle_` from an array to one single instance so fwd / bwd weight / bwd data in a single layer share the same MIOpen handle.

**Comment by dagamayank** [comment]
@whchung You might want to review these two files if you have not done so already -
https://github.com/NVIDIA/caffe/blob/caffe-0.16/src/caffe/layers/cudnn_conv_layer.cu
https://github.com/NVIDIA/caffe/blob/caffe-0.16/src/caffe/layers/cudnn_conv_layer.cpp

This is what Baidu is running and NOT BVLC Caffe off which we based our implementation on.
I do see they have a way to share the workspace, for example, and also use one handle.

**Comment by whchung** [comment]
@dagamayank thanks. I can see they get rid of potentially dangerous `new`/`delete`, in my other branch `ex_no_new_in_conv` I'm verifying the same thing.

# of MIOpen handles definitely need to be reduced.

**Comment by whchung** [comment]
@sunway513 conflicts resolved

**Pull request closed without merging**